tests/regressions/lcos/future_2667.cpp

```diff
@@ -0,0 +1,40 @@
+//  Copyright (c) 2017 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+// This test case demonstrates the issue described in #2667: Ambiguity of
+// nested hpx::future<void>'s
+
+#include <hpx/hpx.hpp>
+#include <hpx/hpx_main.hpp>
+#include <hpx/util/lightweight_test.hpp>
+
+#include <chrono>
+#include <utility>
+
+void do_more_work()
+{
+    hpx::this_thread::sleep_for(std::chrono::seconds(1));
+}
+
+int main()
+{
+    hpx::future<hpx::future<void> > fut =
+        hpx::async(
+            []() -> hpx::future<void> {
+                return hpx::async(
+                    []() -> void {
+                        do_more_work();
+                    });
+            });
+
+    hpx::util::high_resolution_timer t;
+
+    hpx::future<void> fut2 = std::move(fut);
+    fut2.get();
+
+    HPX_TEST(t.elapsed() > 1.0);
```

Comment:
Yes, checking a flag can be done additionally. The time check is supposed to make sure that the collapsed future becomes ready only after the inner future.